File: app/src/main/java/br/com/jonathanzanella/myexpenses/receipt/ReceiptAdapter.java
package br.com.jonathanzanella.myexpenses.receipt;

import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.joda.time.DateTime;

import java.util.List;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.account.Account;
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl;
import br.com.jonathanzanella.myexpenses.helpers.AdapterColorHelper;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper;
import br.com.jonathanzanella.myexpenses.source.Source;
import butterknife.Bind;
import butterknife.ButterKnife;

class ReceiptAdapter extends RecyclerView.Adapter<ReceiptAdapter.ViewHolder> {
	protected List<Receipt> receipts;
	private final ReceiptAdapterPresenter presenter;
	private DateTime date;

	public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
		@Bind(R.id.row_receipt_name)
		TextView name;
		@Bind(R.id.row_receipt_date)
		TextView date;
		@Bind(R.id.row_receipt_income)
		TextView income;
		@Bind(R.id.row_receipt_source)
		TextView source;
		@Bind(R.id.row_receipt_account)
		TextView account;
		@Bind(R.id.row_receipt_show_in_resume_stt)
		TextView showInResume;

		private final AdapterColorHelper adapterColorHelper;

		public ViewHolder(View itemView) {
			super(itemView);

			ButterKnife.bind(this, itemView);

			//noinspection deprecation
			int oddColor = itemView.getContext().getResources().getColor(R.color.color_list_odd);
			//noinspection deprecation
			int evenColor = itemView.getContext().getResources().getColor(R.color.color_list_even);
			adapterColorHelper = new AdapterColorHelper(oddColor, evenColor);

			itemView.setOnClickListener(this);
		}

		@UiThread
		public void setData(final Receipt receipt) {
			itemView.setTag(receipt.getUuid());
			itemView.setBackgroundColor(adapterColorHelper.getColorForGridWithTwoColumns(getAdapterPosition()));
			name.setText(receipt.getName());
			date.setText(Receipt.SIMPLE_DATE_FORMAT.format(receipt.getDate().toDate()));
			income.setText(CurrencyHelper.format(receipt.getIncome()));

			new AsyncTask<Void, Void, Source>() {

				@Override
				protected Source doInBackground(Void... voids) {
					return receipt.getSource();
				}

				@Override
				protected void onPostExecute(Source s) {
					super.onPostExecute(s);
					source.setText(s.getName());
				}
			}.execute();

			new AsyncTask<Void, Void, Account>() {

				@Override
				protected Account doInBackground(Void... voids) {
					return receipt.getAccountFromCache();
				}

				@Override
				protected void onPostExecute(Account a) {
					super.onPostExecute(a);
					account.setText(a.getName());
				}
			}.execute();

			showInResume.setVisibility(receipt.isShowInResume() ? View.VISIBLE : View.INVISIBLE);
		}

		@Override
		public void onClick(View v) {
			Receipt receipt = getReceipt(getAdapterPosition());
			if(receipt != null) {
                Intent i = new Intent(itemView.getContext(), ShowReceiptActivity.class);
                i.putExtra(ShowReceiptActivity.Companion.getKEY_RECEIPT_UUID(), receipt.getUuid());
                itemView.getContext().startActivity(i);
			}
		}
	}

	ReceiptAdapter(Context context) {
		presenter = new ReceiptAdapterPresenter(new ReceiptRepository(new RepositoryImpl<Receipt>(context)));
	}

	public void loadData(DateTime date) {
		receipts = presenter.getReceipts(true, date);
		this.date = date;
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_receipt, parent, false);
		return new ViewHolder(v);
	}

	@Override
	public void onBindViewHolder(ViewHolder holder, int position) {
		holder.setData(getReceipt(position));
	}

	@Override
	public int getItemCount() {
		return receipts != null ? receipts.size() : 0;
	}

	@Nullable
	private Receipt getReceipt(int position) {
		return receipts != null ? receipts.get(position) : null;
	}

	public void filter(String filter) {
		presenter.filter(filter);
		receipts = presenter.getReceipts(false, date);
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/receipt/ReceiptAdapter.kt
package br.com.jonathanzanella.myexpenses.receipt

import android.content.Context
import android.content.Intent
import android.os.AsyncTask
import android.support.annotation.UiThread
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.account.Account
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl
import br.com.jonathanzanella.myexpenses.helpers.AdapterColorHelper
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper
import br.com.jonathanzanella.myexpenses.source.Source
import br.com.jonathanzanella.myexpenses.transaction.Transaction
import br.com.jonathanzanella.myexpenses.views.anko.*
import org.jetbrains.anko.*
import org.joda.time.DateTime

open class ReceiptAdapter(context: Context) : RecyclerView.Adapter<ReceiptAdapter.ViewHolder>() {
    private var receipts: List<Receipt> = ArrayList()
    private val presenter: ReceiptAdapterPresenter = ReceiptAdapterPresenter(ReceiptRepository(RepositoryImpl<Receipt>(context)))
    private var date: DateTime? = null

    inner class ViewHolder internal constructor(itemView: View, private val ui: ViewUI) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private val adapterColorHelper: AdapterColorHelper

        init {
            val oddColor = ResourcesCompat.getColor(itemView.context.resources, R.color.color_list_odd, null)
            val evenColor = ResourcesCompat.getColor(itemView.context.resources, R.color.color_list_even, null)
            adapterColorHelper = AdapterColorHelper(oddColor, evenColor)

            itemView.setOnClickListener(this)
        }

        @UiThread
        fun setData(receipt: Receipt) {
            itemView.tag = receipt.uuid
            itemView.setBackgroundColor(adapterColorHelper.getColorForGridWithTwoColumns(adapterPosition))
            ui.name.text = receipt.name
            ui.date.text = Transaction.SIMPLE_DATE_FORMAT.format(receipt.getDate().toDate())
            ui.income.text = CurrencyHelper.format(receipt.income)

            object : AsyncTask<Void, Void, Source>() {

                override fun doInBackground(vararg voids: Void): Source {
                    return receipt.source
                }

                override fun onPostExecute(s: Source) {
                    super.onPostExecute(s)
                    ui.source.text = s.name
                }
            }.execute()

            object : AsyncTask<Void, Void, Account>() {

                override fun doInBackground(vararg voids: Void): Account {
                    return receipt.accountFromCache
                }

                override fun onPostExecute(a: Account) {
                    super.onPostExecute(a)
                    ui.account.text = a.name
                }
            }.execute()

            ui.showInResume.visibility = if (receipt.isShowInResume) View.VISIBLE else View.INVISIBLE
        }

        override fun onClick(v: View) {
            val i = Intent(itemView.context, ShowReceiptActivity::class.java)
            i.putExtra(ShowReceiptActivity.KEY_RECEIPT_UUID, getReceipt(adapterPosition).uuid)
            itemView.context.startActivity(i)
        }
    }

    fun loadData(date: DateTime) {
        receipts = presenter.getReceipts(true, date)
        this.date = date
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val ui = ViewUI()
        return ViewHolder(ui.createView(AnkoContext.create(parent.context, parent)), ui)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setData(getReceipt(position))
    }

    override fun getItemCount(): Int {
        return receipts.size
    }

    private fun getReceipt(position: Int): Receipt {
        return receipts[position]
    }

    fun filter(filter: String) {
        presenter.filter(filter)
        receipts = presenter.getReceipts(false, date)
    }
}

internal class ViewUI: AnkoComponent<ViewGroup> {
    lateinit var name: TextView
    lateinit var date: TextView
    lateinit var income: TextView
    lateinit var source: TextView
    lateinit var account: TextView
    lateinit var showInResume: TextView

    override fun createView(ui: AnkoContext<ViewGroup>) = with(ui) {
        relativeLayout {
            padding = resources.getDimensionPixelSize(R.dimen.row_spacing)

            name = rowPrincipalInformation { id = R.id.row_receipt_name
            }.lparams {
                marginEnd = resources.getDimensionPixelSize(R.dimen.min_spacing)
                leftOf(R.id.row_receipt_income)
                alignParentStart()
            }
            date = rowDetailInformation { id = R.id.row_receipt_date }.lparams { below(R.id.row_receipt_name) }
            income = rowSecondaryInformation { id = R.id.row_receipt_income }.lparams { alignParentEnd() }

            linearLayout {
                id = R.id.row_receipt_source_layout
                orientation = LinearLayout.HORIZONTAL

                rowStaticInformation { text = resources.getString(R.string.source)
                }.lparams { marginEnd = resources.getDimensionPixelSize(R.dimen.min_spacing) }

                source = rowSecondaryInformation { id = R.id.row_receipt_source }
            }.lparams(height = wrapContent) {
                topMargin = resources.getDimensionPixelSize(R.dimen.min_spacing)
                below(R.id.row_receipt_date)
            }

            linearLayout {
                id = R.id.row_receipt_account_layout
                orientation = LinearLayout.HORIZONTAL

                rowStaticInformation { text = resources.getString(R.string.account)
                }.lparams { marginEnd = resources.getDimensionPixelSize(R.dimen.min_spacing) }

                account = rowSecondaryInformation { id = R.id.row_receipt_account }
            }.lparams(height = wrapContent) {
                topMargin = resources.getDimensionPixelSize(R.dimen.min_spacing)
                below(R.id.row_receipt_source_layout)
            }

            showInResume = rowDetailInformation {
                id = R.id.row_receipt_show_in_resume_stt
                text = resources.getString(R.string.show_in_resume)
                visibility = View.INVISIBLE
            }.lparams {
                below(R.id.row_receipt_account_layout)
                alignParentEnd()
            }
        }.applyRecursively(::applyTemplateViewStyles)
    }
}